### PR TITLE
userns: prevent /sys/kernel/* paths in the container

### PR DESCRIPTION
**Pull request by giuseppe**
when we run in a user namespace, there are cases where we have not
enough privileges to mount a fresh sysfs on /sys.  To circumvent this
limitation, we rbind /sys from the host.  This carries inside of the
container also some mounts we probably don't want to.  We are also
limited by the kernel to use rbind instead of bind, as allowing a bind
would uncover paths that were not previously visible.

This is a slimmed down version of the intermediate mount namespace
logic we had before, where we only set /sys to slave, so the umounts
done to the storage by the cleanup process are propagated back to the
host.  We also don't setup any new directory, so there is no
additional cleanup to do.

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

### CODE REVIEW

File: libpod/oci_linux.go [hunk 1]
```diff
@@ -91,6 +94,53 @@ func (r *OCIRuntime) createContainer(ctr *Container, cgroupParent string, restor
...
+
+					// don't spread our mounts around.  We are setting only /sys to be slave
+					// so that the cleanup process is still able to umount the storage and the
+					// changes are propagated to the host.
+					err = unix.Mount("/sys", "/sys", "none", unix.MS_REC|unix.MS_SLAVE, "")
+					if err != nil {
+						return errors.Wrapf(err, "cannot make /sys slave")
+					}
+
+					paths := []string{
```
Comment (rhatdan) [inline libpod/oci_linux.go]: Could we make this more future proof.  Remove all submounts of /sys?  Or at least anything begining with /sys/kernel?

File: test/system/030-run.bats [hunk 2]
```diff
@@ -31,4 +31,12 @@ echo $rand        |   0 | $rand
     done < <(parse_table "$tests")
 }
 
+@test "podman run - uidmapping has no /sys/kernel mounts" {
+      run_podman $expected_rc run --uidmapping 0:100:10000 $IMAGE mount | grep /sys/kernel
```
Comment (rhatdan) [inline test/system/030-run.bats]: I think you need --net=host here to make the current behaviour happen.

Comment (giuseppe) [reply test/system/030-run.bats]: I've added another test with --net=host, so that we can test both cases

File: libpod/oci_linux.go (hunk 1 above)
Comment (giuseppe) [reply libpod/oci_linux.go]: fixed now

**Comment by rhatdan** [comment]
LGTM
@TomSweeneyRedHat PTAL
@vrothberg @mheon @baude PTAL

**Comment by vrothberg** [comment]
@giuseppe rootless tests aren't happy

**Comment by giuseppe** [comment]
ah, this wouldn't work with rootless.  An unprivileged user cannot umount these paths.  @rhatdan is it an issue for buildah?

File: libpod/oci_linux.go [hunk 3]
```diff
@@ -91,6 +96,55 @@ func (r *OCIRuntime) createContainer(ctr *Container, cgroupParent string, restor
...
+						return err
+					}
+					defer fd.Close()
+					defer unix.Setns(int(fd.Fd()), unix.CLONE_NEWNS)
+
+					// create a new mountns on the current thread
+					if err = unix.Unshare(unix.CLONE_NEWNS); err != nil {
+						return err
+					}
+					defer unix.Setns(int(fd.Fd()), unix.CLONE_NEWNS)
```
Comment (TomSweeneyRedHat) [inline libpod/oci_linux.go]: is this defer needed?  Seems to be a duplicate of line 112

**Comment by rhatdan** [comment]
@giuseppe Yes.
podman run --net=host fedora mount | grep /sys/kernel
securityfs on /sys/kernel/security type securityfs (rw,nosuid,nodev,noexec,relatime)
configfs on /sys/kernel/config type configfs (rw,relatime)
debugfs on /sys/kernel/debug type debugfs (rw,relatime,seclabel)

File: libpod/oci_linux.go (hunk 3 above)
Comment (giuseppe) [reply libpod/oci_linux.go]: good catch!  Fixed

**Comment by rhatdan** [comment]
We might need to mask these in rootless mode.  I am not sure if @TomSweeneyRedHat has tried buildah bud --isolation chroot inside of rootless podman yet.

We really need this for root running container anyways, especially where if you don't have SELinux running you can potentially modify the kernel.

In rootless mode this would obviously be blocked.

**Comment by giuseppe** [comment]
> @giuseppe Yes.
> podman run --net=host fedora mount | grep /sys/kernel
> securityfs on /sys/kernel/security type securityfs (rw,nosuid,nodev,noexec,relatime)
> configfs on /sys/kernel/config type configfs (rw,relatime)
> debugfs on /sys/kernel/debug type debugfs (rw,relatime,seclabel)

I think there is no way for an unprivileged user to drop that mounts.  It cannot mount a new sysfs and it cannot umount anything that will reveal what is under the mount point.  On the other hand, is it really a security issue?  The unprivileged user won't have any new privilege that it didn't have on the host.

File: libpod/oci_linux.go (hunk 3 above)
Comment (rhatdan) [reply libpod/oci_linux.go]: I think we don't need the one above.

**Comment by rhatdan** [comment]
Right I am worried about information reveal though.  The information in /sys/kernel/debug might be useful to a hacked container.  If we can not umount we should mount a tmpfs over them.

**Comment by giuseppe** [comment]
> Right I am worried about information reveal though. The information in /sys/kernel/debug might be useful to a hacked container. If we can not umount we should mount a tmpfs over them.

we can mask it, altough `/sys/kernel/debug` has mode 0700, an unprivileged user cannot access it

**Comment by giuseppe** [comment]
> Right I am worried about information reveal though. The information in /sys/kernel/debug might be useful to a hacked container. If we can not umount we should mount a tmpfs over them.

I've added a patch to mask `/sys/kernel` if we are bind mounting `/sys` and we are rootless

**Comment by rhatdan** [comment]
Well this looks good.
find /sys/kernel/ | wc
find: '/sys/kernel/security': Permission denied
find: '/sys/kernel/debug': Permission denied
find: '/sys/kernel/config': Permission denied
5615 5615 213373
So I think for rootless we can ignore the mounts.

**Comment by giuseppe** [comment]
> Well this looks good.
> find /sys/kernel/ | wc
> find: '/sys/kernel/security': Permission denied
> find: '/sys/kernel/debug': Permission denied
> find: '/sys/kernel/config': Permission denied
> 5615 5615 213373
> So I think for rootless we can ignore the mounts.

should I drop the second patch or is it still a valid additional defense?

**Comment by mheon** [comment]
LGTM. I think we keep the masked path - defense in depth is good.

**Comment by mheon** [comment]
Holding /lgtm until tests go green

**Comment by edsantiago** [comment]
This commit seems to have broken the new unbrivileged-access BATS test:
```
# bats test/system/400-unprivileged-access.bats
 ✗ podman container storage is not accessible by unprivileged users
   (in test file test/system/400-unprivileged-access.bats, line 74)
     `find $GRAPH_ROOT $RUN_ROOT \! -type l -perm -o+w -print | while read i; do' failed
   $ /home/esm/src/atomic/2018-02.podman/libpod/bin/podman rm --all --force
   $ /home/esm/src/atomic/2018-02.podman/libpod/bin/podman images --all --format {{.Repository}}:{{.Tag}} {{.ID}}
   quay.io/libpod/alpine_labels:latest   4fab981df737
   $ /home/esm/src/atomic/2018-02.podman/libpod/bin/podman run --name c_uidmap --uidmap 0:10000:10000 quay.io/libpod/alpine_labels:latest true
   $ /home/esm/src/atomic/2018-02.podman/libpod/bin/podman run --name c_uidmap_v --uidmap 0:10000:10000 -v foo:/foo quay.io/libpod/alpine_labels:latest true
   $ /home/esm/src/atomic/2018-02.podman/libpod/bin/podman run --name c_mount quay.io/libpod/alpine_labels:latest sh -c echo hi > /myfile;mkdir -p /mydir/mysubdir; chmod 777 /myfile /mydir /mydir/mysubdir
   $ /home/esm/src/atomic/2018-02.podman/libpod/bin/podman mount c_mount
   /var/lib/containers/storage/overlay/1440f9d7c7b5a8fc4bb1f4d699ef156960f89980663357b2481258ca7430b3d1/merged
   $ /home/esm/src/atomic/2018-02.podman/libpod/bin/podman info --format {{.store.GraphRoot}}
   /var/lib/containers/storage
   $ /home/esm/src/atomic/2018-02.podman/libpod/bin/podman info --format {{.store.RunRoot}}
   /var/run/containers/storage
   chmod: changing permissions of '/var/lib/containers/storage/overlay/71071cf29c423da6351954badf4aae660aa3ca1ebcdb8fedce8fda7c659b4599/merged': Operation not permitted
   chmod: changing permissions of '/var/lib/containers/storage/overlay/71071cf29c423da6351954badf4aae660aa3ca1ebcdb8fedce8fda7c659b4599/merged/tmp': Operation not permitted
   #/vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv
   #| FAIL: Able to run 'ls /var/lib/containers/storage/overlay/71071cf29c423da6351954badf4aae660aa3ca1ebcdb8fedce8fda7c659b4599/merged/tmp' without error
   #\^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
@giuseppe any suggestions on how to fix it

**Comment by giuseppe** [comment]
> This commit seems to have broken the new unbrivileged-access BATS test:

I think it is related to the new tests that are using a mapping not correct for the unprivileged access tests

**Comment by giuseppe** [comment]
I am surprised how that could happen though, directories under `/var/lib/containers/storage/overlay` have mode `drwx------` that blocks completely access to other users

**Comment by edsantiago** [comment]
Not any more. Immediately after this command, `/var/lib/containers/storage` becomes 711:
```
# ./bin/podman run --name c_uidmap   --uidmap 0:10000:10000 alpine_labels true 
```
Furthermore, I am now stuck in a system where I can no longer remove that directory:
```
# rm -rf /var/lib/containers/storage
rm: cannot remove '/var/lib/containers/storage/overlay/0edc5e399bf04596f445dfaf7073d15699e2327dc45e1f62dca43638fc4095b8/merged': Device or resource busy
```
(podman rm -a, podman ps -a, podman images -a, all show nothing).

master @ c1e2b583c82261639d6927c9b2a4f19d87a166a7. I don't have time this evening to do a full investigateion but wanted to leave you with something you can check out tomorrow morning.

**Comment by giuseppe** [comment]
how do directories under `/var/lib/containers/storage/overlay` look like?

```
$ sudo ls -l /var/lib/containers/storage/overlay
total 16
drwx------. 5 root root  4096 Apr 15 23:39 3f72eff9ab5617186824e3a431489b5bb70fadc3004734b3c66008da3f3f0724
drwx------. 6 root root  4096 Apr 15 23:26 a464c54f93a9e88fc1d33df1e0e39cca427d60145a360962e8f19a1dbf900da9
drwx------. 7  100 users 4096 Apr 15 23:11 df64d3292fd6194b7865d7326af5255db6d81e9df29f48adde61a918fbd8c332
drwx------. 2 root root  4096 Apr 16 09:39 l
```

They must have mode `drwx------`.

**Comment by edsantiago** [comment]
Sorry for the late reply. I can no longer reproduce this on my system, and OSP10 is down so I can't get a new virt for testing. The only difference on my system is that I'm still stuck with the undeletable directory mentioned above, so instead of `rm -rf /var/lib/containers/storage` I've done `mv ...{,.DELETE-ME}`. There is otherwise no difference: no reboot, no dnf-upgrade. I've tried everything I can to reproduce.

**Pull request merged**